cmd/contour/cli.go

```diff
@@ -118,21 +140,242 @@ type stream interface {
 	Recv() (*envoy_discovery_v3.DiscoveryResponse, error)
 }
 
-func watchstream(st stream, typeURL string, resources []string) {
-	m := proto.TextMarshaler{
-		Compact:   false,
-		ExpandAny: true,
+func watchstream(log *logrus.Logger, st stream, typeURL string, resources []string, nack bool, nodeID string) {
+	m := protojson.MarshalOptions{
+		Multiline:     true,
+		Indent:        "  ",
+		UseProtoNames: true,
+	}
+
+	currentVersion := "0"
+
+	// Send the initial, non-ACK discovery request.
+	req := &envoy_discovery_v3.DiscoveryRequest{
+		TypeUrl:       typeURL,
+		ResourceNames: resources,
+		VersionInfo:   currentVersion,
+		Node: &corev3.Node{
+			Id: nodeID,
+		},
 	}
+	log.WithField("currentVersion", currentVersion).Info("Sending discover request")
+	fmt.Println(m.Format(req))
+	err := st.Send(req)
+	if err != nil {
+		log.WithError(err).Fatal("failed to send Discover Request")
+	}
+
 	for {
-		req := &envoy_discovery_v3.DiscoveryRequest{
-			TypeUrl:       typeURL,
-			ResourceNames: resources,
+
+		// Wait until we receive a response to our request.
+		resp, err := st.Recv()
+		if err != nil {
+			log.WithError(err).Fatal("failed to receive response for Discover Request")
+		}
+		log.WithField("currentVersion", currentVersion).
+			WithField("resp_version_info", resp.VersionInfo).
+			WithField("nonce", resp.Nonce).
+			Info("Received Discovery Response")
+
+		fmt.Println(m.Format(resp))
+		if err != nil {
+			log.WithError(err).Fatal("failed to marshal Discovery Response")
 		}
-		err := st.Send(req)
-		kingpin.FatalIfError(err, "failed to send Discover Request")
+
+		currentVersion = resp.VersionInfo
+
+		if nack {
+			// We'll NACK the response we just got.
+			// The ResponseNonce field is what makes it an ACK,
+			// and the VersionInfo field must match the one in the response we
```

Comment:
I think the version info in a NACK or ACK should match the last good response, so in this NACK case should still return an empty version: https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#nack

its a little funny in the NACK case for the CLI to send NACKs forever and starting with the first request, maybe we could add a "NACK sentinel" file so we can test turning NACKs on and off?

Reply:
The NACK sentinel file idea is a good idea, we could also take a keyboard input maybe? Then we can toggle on the NACK, and toggle it off again.

I'll fix the version, but I think we should merge this in, then come back and add in some sort of runtime NACK switch.

Reply:
If we remove sending the version here, go-control-plane will respond to the NACK by immediately sending a DiscoveryResponse, which will mean that we'll spam NACKs and responses back and forth as fast as the wire can handle.

I think we need to think of the VersionInfo as "the last config *seen*", not "the last config *ACKed*", at least until we add better NACK handling to go-control-plane.